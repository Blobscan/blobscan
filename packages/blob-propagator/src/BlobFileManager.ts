import fs from "fs";
import path from "path";

import type { Blob } from "./types";

export type BlobFileManagerConfig = {
  basePath: string;
  folderName: string;
};

export class BlobFileManager {
  #blobFolderPath: string;

  constructor({
    basePath = __dirname,
    folderName = "eip4844-blobs",
  }: Partial<BlobFileManagerConfig> = {}) {
    this.#blobFolderPath = path.join(basePath, folderName);

    if (!fs.existsSync(this.#blobFolderPath)) {
      fs.mkdirSync(this.#blobFolderPath);
    }
  }

  checkFileExists(versionedHash: string) {
    const path = this.#buildFilePath(versionedHash);

    return fs.promises
      .access(path, fs.constants.F_OK)
      .then(() => true)
      .catch(() => false);
  }

  async createFile({ data, versionedHash }: Blob) {
    try {
      if (await this.checkFileExists(versionedHash)) {
        return Promise.resolve();
      }

      const blobfilePath = this.#buildFilePath(versionedHash);

      return fs.promises.writeFile(blobfilePath, data, { encoding: "utf-8" });
    } catch (err) {
      throw new Error(
        `Couldn't store blob ${versionedHash} data on filesystem: ${err}`
      );
    }
  }

  async readFile(versionedHash: string) {
    const blobFilePath = this.#buildFilePath(versionedHash);

    try {
      const blobData = await fs.promises.readFile(blobFilePath, "utf-8");

      return blobData;
    } catch (error) {
      throw new Error(
        `Couldn't read blob ${versionedHash} data file: file is missing`
      );
    }
  }

  async removeFile(versionedHash: string) {
    try {
      await fs.promises.unlink(this.#buildFilePath(versionedHash));
    } catch (err) {
      throw new Error(
        `Couldn't remove blob ${versionedHash} data file: ${err}`
      );
    }
  }

  async removeFolder() {
    return fs.promises.rm(this.#blobFolderPath, {
      recursive: true,
    });
  }

  #buildFilePath(versionedHash: string) {
    return path.join(this.#blobFolderPath, `${versionedHash}.txt`);
  }
}
