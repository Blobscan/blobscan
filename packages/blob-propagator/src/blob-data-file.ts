import fs from "fs";

import type { Blob } from "./types";

export function fileExists(path: string) {
  return fs.promises
    .access(path, fs.constants.F_OK)
    .then(() => true)
    .catch(() => false);
}

export function buildBlobDataFilePath(versionedHash: string) {
  return `${versionedHash}.txt`;
}

export async function createBlobDataFile({ data, versionedHash }: Blob) {
  const blobfilePath = buildBlobDataFilePath(versionedHash);

  try {
    if (await fileExists(blobfilePath)) {
      return Promise.resolve();
    }

    return fs.promises.writeFile(blobfilePath, data, { encoding: "utf-8" });
  } catch (err) {
    throw new Error(
      `Couldn't store blob ${versionedHash} data on filesystem: ${err}`
    );
  }
}

export async function readBlobDataFile(versionedHash: string) {
  const blobFilePath = buildBlobDataFilePath(versionedHash);

  try {
    const blobData = await fs.promises.readFile(blobFilePath, "utf-8");

    return blobData;
  } catch (error) {
    throw new Error(
      `Couldn't read blob ${versionedHash} data file: file is missing`
    );
  }
}

export async function removeBlobDataFile(versionedHash: string) {
  try {
    await fs.promises.unlink(buildBlobDataFilePath(versionedHash));
  } catch (err) {
    throw new Error(`Couldn't remove blob ${versionedHash} data file: ${err}`);
  }
}
