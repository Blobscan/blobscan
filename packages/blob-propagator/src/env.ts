import {
  booleanSchema,
  createEnv,
  maskPassword,
  presetEnvOptions,
  z,
} from "@blobscan/zod";

export const env = createEnv({
  envOptions: {
    server: {
      REDIS_URI: z.string().default("redis://localhost:6379"),

      POSTGRES_STORAGE_ENABLED: booleanSchema.default("false"),
      GOOGLE_STORAGE_ENABLED: booleanSchema.default("false"),
      SWARM_STORAGE_ENABLED: booleanSchema.default("false"),

      BLOB_PROPAGATOR_ENABLED: booleanSchema.default("false"),
      TEST: booleanSchema.optional(),
    },

    ...presetEnvOptions,
  },
  display(env) {
    console.log(
      `Blob propagator configuration: enabled=${
        env.BLOB_PROPAGATOR_ENABLED
      } redisUri=${maskPassword(env.REDIS_URI)}`
    );
  },
});

export type EnvVars = typeof env;
