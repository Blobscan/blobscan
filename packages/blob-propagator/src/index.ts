export type { EnvVars as Environment } from "./env";

export { env } from "./env";
export { BlobPropagator } from "./BlobPropagator";
export { blobPropagator } from "./blob-propagator";
export * from "./types";
export {
  createBlobPropagationFlowJob,
  createBlobStorageJob,
  FINALIZER_WORKER_NAME,
  STORAGE_WORKER_NAMES,
  buildJobId,
} from "./utils";
