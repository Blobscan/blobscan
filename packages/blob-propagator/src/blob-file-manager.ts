import os from "os";

import { BlobFileManager } from "./BlobFileManager";
import { env } from "./env";

const basePath = !env.TEST ? os.tmpdir() : undefined;

export const blobFileManager = new BlobFileManager({
  basePath,
  folderName: "blobscan-blobs",
});
