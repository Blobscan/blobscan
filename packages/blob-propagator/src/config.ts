import type { QueueOptions, WorkerOptions } from "bullmq";

import type { BlobStorage } from "@blobscan/db";

import { env } from "./env";

export const STORAGE_QUEUES: Record<BlobStorage, string> = {
  GOOGLE: "gcs-queue",
  SWARM: "swarm-queue",
  POSTGRES: "postgres-queue",
};

export const FINALIZER_QUEUE = "finalizer-queue";

export const DEFAULT_OPTIONS: QueueOptions = {
  defaultJobOptions: {
    backoff: {
      type: "exponential",
    },
    attempts: 3,
    removeOnComplete: true,
  },
};

export const DEFAULT_WORKER_OPTIONS: WorkerOptions = {
  useWorkerThreads: true,
  connection: {
    host: env.REDIS_QUEUE_HOST,
    port: env.REDIS_QUEUE_PORT,
    password: env.REDIS_QUEUE_PASSWORD,
    username: env.REDIS_QUEUE_USERNAME,
  },
};
