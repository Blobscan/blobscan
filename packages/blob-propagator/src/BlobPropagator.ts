/* eslint-disable @typescript-eslint/no-misused-promises */
import { FlowProducer, Queue, Worker } from "bullmq";
import type { ConnectionOptions, FlowJob, WorkerOptions } from "bullmq";

import type { $Enums } from "@blobscan/db";
import type { BlobStorage } from "@blobscan/db";
import { logger } from "@blobscan/logger";

import { blobFileManager } from "./blob-file-manager";
import type { Blob, BlobPropagationJobData } from "./types";
import {
  FINALIZER_WORKER_NAME,
  STORAGE_WORKER_NAMES,
  createBlobPropagationFlowJob,
} from "./utils";
import {
  fileSystemProcessor,
  finalizerProcessor,
  gcsProcessor,
  postgresProcessor,
  swarmProcessor,
} from "./worker-processors";

const STORAGE_WORKER_PROCESSORS = {
  GOOGLE: gcsProcessor,
  SWARM: swarmProcessor,
  POSTGRES: postgresProcessor,
  FILE_SYSTEM: fileSystemProcessor,
};

export type BlobPropagatorOptions = Partial<{
  workerOptions: WorkerOptions;
}>;

const DEFAULT_WORKER_OPTIONS: WorkerOptions = {
  autorun: true,
  useWorkerThreads: false,
  removeOnComplete: { count: 1000 },
};

export class BlobPropagator {
  protected blobPropagationFlowProducer: FlowProducer;
  protected finalizerWorker: Worker;
  protected storageWorkers: Record<
    $Enums.BlobStorage,
    Worker<BlobPropagationJobData>
  >;

  constructor(
    storages: $Enums.BlobStorage[],
    { workerOptions }: BlobPropagatorOptions = {}
  ) {
    if (!storages.length) {
      throw new Error(
        "Couldn't instantiate blob propagator: no storages given"
      );
    }

    this.storageWorkers = this.#createBlobStorageWorkers(
      storages,
      workerOptions
    );
    this.finalizerWorker = this.#createFinalizerWorker(workerOptions);
    this.blobPropagationFlowProducer = this.#createBlobPropagationFlowProducer(
      workerOptions?.connection
    );
  }

  async empty({ force }: { force: boolean } = { force: false }) {
    const workers = this.#getWorkers();
    const queues = await Promise.all(
      workers.map(
        async (w) => new Queue(w.name, { connection: await w.client })
      )
    );

    let emptyPromise = Promise.all([
      ...queues.map((q) => q.obliterate({ force })),
      blobFileManager.removeFolder(),
    ]);

    queues.forEach((q) => {
      emptyPromise = emptyPromise.finally(async () => {
        await q.close();
      });
    });

    await emptyPromise;
  }

  close() {
    let teardownPromise: Promise<void> = Promise.resolve();

    Object.values(this.storageWorkers).forEach((w) => {
      teardownPromise = teardownPromise.finally(async () => {
        await w.close();
      });
    });

    return teardownPromise
      .finally(async () => {
        await this.finalizerWorker.close();
      })
      .finally(async () => {
        const redisClient = await this.blobPropagationFlowProducer.client;

        await redisClient.quit();
      })
      .finally(async () => {
        await this.blobPropagationFlowProducer.close();
      });
  }

  async propagateBlob(blob: Blob) {
    await blobFileManager.createFile(blob);

    const blobFlowJob = await this.#createBlobPropagationFlowJob({
      versionedHash: blob.versionedHash,
    });

    await this.blobPropagationFlowProducer.add(blobFlowJob);
  }

  async propagateBlobs(blobs: Blob[]) {
    const uniqueBlobs = Array.from(
      new Set(blobs.map((b) => b.versionedHash))
    ).map((versionedHash) => {
      // @typescript-eslint/no-non-null-assertion
      const blob = blobs.find((b) => b.versionedHash === versionedHash)!;

      return blob;
    });

    await Promise.all(
      uniqueBlobs.map((blob) => blobFileManager.createFile(blob))
    );

    const blobPropagationFlowJobs = uniqueBlobs.map(({ versionedHash }) =>
      this.#createBlobPropagationFlowJob({ versionedHash })
    );

    await this.blobPropagationFlowProducer.addBulk(blobPropagationFlowJobs);
  }

  #getWorkers() {
    return [...Object.values(this.storageWorkers), this.finalizerWorker];
  }

  #createBlobPropagationFlowProducer(connection?: ConnectionOptions) {
    /*
     * Instantiating a new `FlowProducer` appears to create two separate `RedisConnection` instances.
     * This leads to an issue where one instance remains active, or "dangling", after the `FlowProducer` has been closed.
     * To prevent this, we now initialize a single `IORedis` instance in advance and use it when creating the `FlowProducer`.
     * This way, both created connections reference the same `IORedis` instance and can be closed properly.
     *
     * See: https://github.com/taskforcesh/bullmq/blob/d7cf6ea60830b69b636648238a51e5f981616d02/src/classes/flow-producer.ts#L111
     */
    const blobPropagationFlowProducer = new FlowProducer({
      connection,
    });

    blobPropagationFlowProducer.on("error", (err) => {
      logger.error(`Blob propagation flow producer error: ${err}`);
    });

    return blobPropagationFlowProducer;
  }

  #createFinalizerWorker(opts: WorkerOptions = {}) {
    const finalizerWorker = new Worker(
      FINALIZER_WORKER_NAME,
      finalizerProcessor,
      {
        ...DEFAULT_WORKER_OPTIONS,
        ...opts,
      }
    );

    finalizerWorker.on("completed", (job) => {
      logger.debug(`Job ${job.id} completed`);
    });

    finalizerWorker.on("failed", (job, err) => {
      logger.error(`Job ${job?.id} failed: ${err}`);
    });

    return finalizerWorker;
  }

  #createBlobStorageWorkers(
    storages: $Enums.BlobStorage[],
    opts: WorkerOptions = {}
  ) {
    return storages.reduce<Record<BlobStorage, Worker<BlobPropagationJobData>>>(
      (workers, storageName) => {
        const workerName = STORAGE_WORKER_NAMES[storageName];
        const storageWorker = new Worker<BlobPropagationJobData>(
          workerName,
          STORAGE_WORKER_PROCESSORS[storageName],
          {
            ...DEFAULT_WORKER_OPTIONS,
            ...opts,
          }
        );

        storageWorker.on("completed", (job) => {
          logger.debug(`Job ${job.id} completed by ${workerName}`);
        });

        storageWorker.on("failed", (job, err) => {
          logger.error(`Job ${job?.id} failed: ${err} (worker: ${workerName})`);
        });

        workers[storageName] = storageWorker;

        return workers;
      },
      {} as Record<BlobStorage, Worker<BlobPropagationJobData>>
    );
  }

  #createBlobPropagationFlowJob(data: BlobPropagationJobData): FlowJob {
    const versionedHash = data.versionedHash;

    const storageWorkerNames = Object.values(this.storageWorkers).map(
      (storageWorker) => storageWorker.name
    );

    return createBlobPropagationFlowJob(
      this.finalizerWorker.name,
      storageWorkerNames,
      versionedHash
    );
  }
}
