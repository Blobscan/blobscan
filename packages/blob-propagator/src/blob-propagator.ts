import IORedis from "ioredis";

import { BlobStorage } from "@blobscan/db";

import { BlobPropagator } from "./BlobPropagator";
import { env } from "./env";

function createBlobPropagator() {
  const availableStorages: BlobStorage[] = [];

  if (env.GOOGLE_STORAGE_ENABLED) {
    availableStorages.push(BlobStorage.GOOGLE);
  }

  if (env.POSTGRES_STORAGE_ENABLED) {
    availableStorages.push(BlobStorage.POSTGRES);
  }

  if (env.SWARM_STORAGE_ENABLED) {
    availableStorages.push(BlobStorage.SWARM);
  }

  const connection = new IORedis(env.REDIS_URI, { maxRetriesPerRequest: null });

  return new BlobPropagator(availableStorages, {
    workerOptions: {
      connection,
    },
  });
}
const blobPropagator =
  env.BLOB_PROPAGATOR_ENABLED === true ? createBlobPropagator() : undefined;

export { blobPropagator, createBlobPropagator };
