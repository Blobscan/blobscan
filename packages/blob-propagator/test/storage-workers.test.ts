import {
  afterAll,
  afterEach,
  beforeAll,
  beforeEach,
  describe,
  expect,
  it,
  vi,
} from "vitest";

import {
  createStorageFromEnv,
  getBlobStorageManager,
} from "@blobscan/blob-storage-manager";
import type {
  BlobStorage,
  BlobStorageManager,
  BlobStorageName,
  SwarmStorage,
} from "@blobscan/blob-storage-manager";
import { prisma } from "@blobscan/db";
import type { BlobscanPrismaClient, Blob as DBBlob } from "@blobscan/db";

import { STORAGE_WORKER_PROCESSORS } from "../src/BlobPropagator";
import { env } from "../src/env";
import type {
  BlobPropagationJob,
  BlobPropagationWorkerParams,
  BlobPropagationWorkerProcessor,
} from "../src/types";

function runWorkerTests(
  storageName: BlobStorageName,
  getWorkerParams: () => Promise<BlobPropagationWorkerParams>,
  blob: { versionedHash: string; data: string; blobUri: string }
) {
  const job = {
    data: {
      versionedHash: blob.versionedHash,
    },
  } as BlobPropagationJob;

  describe(`when running the ${storageName} worker`, () => {
    let storageWorkerProcessor: ReturnType<BlobPropagationWorkerProcessor>;
    let blobStorageManager: BlobStorageManager;
    let prisma: BlobscanPrismaClient;

    beforeEach(async () => {
      const workerParams = await getWorkerParams();

      const storageWorker = STORAGE_WORKER_PROCESSORS[storageName];

      storageWorkerProcessor = storageWorker(workerParams);
      blobStorageManager = workerParams.blobStorageManager;
      prisma = workerParams.prisma;
    });

    it(`propagate the blob to the ${storageName} storage correctly`, async () => {
      await storageWorkerProcessor(job);

      const expectedResult = { storage: storageName, data: blob.data };
      const result = await blobStorageManager.getBlobByReferences({
        reference: blob.blobUri,
        storage: storageName,
      });

      expect(result).toEqual(expectedResult);
    });

    it(`should store the ${storageName} blob storage reference in the db once propagated correctly`, async () => {
      await storageWorkerProcessor(job);

      const blobStorageReference =
        await prisma.blobDataStorageReference.findUnique({
          where: {
            blobHash_blobStorage: {
              blobHash: blob.versionedHash,
              blobStorage: storageName,
            },
          },
        });

      expect(blobStorageReference?.blobHash, "Blob hash mismatch").toBe(
        blob.versionedHash
      );
      expect(
        blobStorageReference?.dataReference,
        "Data reference mismatch"
      ).toBe(blob.blobUri);
      expect(blobStorageReference?.blobStorage, "Storage name mismatch").toBe(
        storageName
      );
    });

    it(`should return the correct blob ${storageName} storage reference`, async () => {
      const blobStorageReference = await storageWorkerProcessor(job);

      expect(blobStorageReference).toEqual({
        storage: storageName,
        reference: blob.blobUri,
      });
    });

    it("should throw an error if the blob data to be propagated wasn't found in any of the other storages", async () => {
      const versionedHash = "missingBlobDataFileVersionedHash";
      const jobWithMissingBlobData = {
        data: {
          versionedHash,
        },
      } as BlobPropagationJob;

      await expect(
        storageWorkerProcessor(jobWithMissingBlobData)
      ).rejects.toThrowErrorMatchingInlineSnapshot(
        '"Failed to propagate blob with hash \\"missingBlobDataFileVersionedHash\\": no blob storage references found to retrieve data from"'
      );
      ``;
    });
  });
}
describe("Storage Workers", () => {
  let tmpBlobStorage: BlobStorage;
  const blobVersionedHash = "test-blob-versioned-hash";
  const blobData = "0x1234abcdeff123456789ab34223a4b2c2e";

  const blob: DBBlob = {
    versionedHash: blobVersionedHash,
    commitment: "test-commitment",
    proof: "test-proof",
    insertedAt: new Date(),
    size: 1,
    updatedAt: new Date(),
    firstBlockNumber: null,
  };

  const mockedSwarmBlobUri = "0xswarm-reference";

  beforeAll(async () => {
    const bsm = await getBlobStorageManager();

    const tmpStorage = await createStorageFromEnv(
      env.BLOB_PROPAGATOR_TMP_BLOB_STORAGE
    );

    bsm.addStorage(tmpStorage);

    tmpBlobStorage = tmpStorage;
  });

  afterAll(() => {
    vi.clearAllMocks();
  });

  beforeEach(async () => {
    await tmpBlobStorage.storeBlob(blobVersionedHash, blobData);
    await prisma.blob.create({
      data: blob,
    });
  });

  afterEach(async () => {
    const blobUri = tmpBlobStorage.getBlobUri(blobVersionedHash);

    if (blobUri) {
      await tmpBlobStorage.removeBlob(blobUri);
    }

    await prisma.blobDataStorageReference.deleteMany({
      where: {
        blobHash: blobVersionedHash,
      },
    });
    await prisma.blob.delete({
      where: {
        versionedHash: blobVersionedHash,
      },
    });
  });

  const storages: { storage: BlobStorageName; uri: string }[] = [
    {
      storage: "POSTGRES",
      uri: blobVersionedHash,
    },
    {
      storage: "GOOGLE",
      uri: "70118930558/st/-b/lo/st-blob-versioned-hash.txt",
    },
    {
      storage: "FILE_SYSTEM",
      uri: "test-blobscan-blobs/70118930558/st/-b/lo/st-blob-versioned-hash.txt",
    },
  ];

  storages.forEach(({ storage, uri }) => {
    runWorkerTests(
      storage,
      async () => {
        const blobStorageManager = await getBlobStorageManager();

        return { blobStorageManager, prisma };
      },
      {
        blobUri: uri,
        data: blobData,
        versionedHash: blobVersionedHash,
      }
    );
  });

  runWorkerTests(
    "SWARM",
    async () => {
      const blobStorageManager = await getBlobStorageManager();

      vi.spyOn(blobStorageManager, "getStorage").mockImplementation(
        (storage) => {
          if (storage === "SWARM") {
            return {
              getBlob: async (uri) => {
                if (uri === mockedSwarmBlobUri) {
                  return Promise.resolve(blobData);
                }

                return Promise.reject(new Error("Blob not found"));
              },
              storeBlob: async (_, __) => {
                return mockedSwarmBlobUri;
              },
            } as SwarmStorage;
          }
        }
      );
      return { blobStorageManager, prisma };
    },
    {
      versionedHash: blobVersionedHash,
      data: blobData,
      blobUri: mockedSwarmBlobUri,
    }
  );
});
