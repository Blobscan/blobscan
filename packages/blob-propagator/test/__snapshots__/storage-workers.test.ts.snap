// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Storage Worker > when running the GOOGLE worker > should throw an error if the blob wasn't found in the primary blob storage 1`] = `"Failed to propagate blob with hash \\"missingBlobDataFileVersionedHash\\""`;

exports[`Storage Worker > when running the GOOGLE worker > should throw an error if the blob wasn't found in the primary blob storage 2`] = `[Error: Failed to retrieve blob from primary storage]`;

exports[`Storage Worker > when running the S3 worker > should throw an error if the blob wasn't found in the primary blob storage 1`] = `"Failed to propagate blob with hash \\"missingBlobDataFileVersionedHash\\""`;

exports[`Storage Worker > when running the S3 worker > should throw an error if the blob wasn't found in the primary blob storage 2`] = `[Error: Failed to retrieve blob from primary storage]`;

exports[`Storage Worker > when running the SWARM worker > should throw an error if the blob wasn't found in the primary blob storage 1`] = `"Failed to propagate blob with hash \\"missingBlobDataFileVersionedHash\\""`;

exports[`Storage Worker > when running the SWARM worker > should throw an error if the blob wasn't found in the primary blob storage 2`] = `[Error: Failed to retrieve blob from primary storage]`;
